std.nu: Rewrite assert method

presidento opened the pull request:
It is a common pattern to add message to assert.
Also the error message was not so helpful. (See the difference in the documentation)

presidento commented:
@amtoine I think with this change even using `assert` to check equality is more helpful than `assert eq`, see the error messages here:

## With assert

```
assert (1 == 1)
assert (10 == 20)
assert (200 == 200)

Error:
  × Assertion failed: No message
    ╭─[myscript:10:1]
 10 │ assert (1 == 1)
 11 │ assert (10 == 20)
    ·         ────┬───
    ·             ╰── It is not true.
 12 │ assert (200 == 200)
    ╰────

```

## With assert eq

```
assert eq 1 1
assert eq 10 20
assert eq 200 200


Error:
  × left is not equal to right
   ╭─[myscript:5:1]
 5 │     if not $cond {
 6 │         error make {msg: $msg}
   ·         ─────┬────
   ·              ╰── originates from here
 7 │     }
   ╰────
```

I'd consider the value added by `assert eq`.

At first step I did not want to make much change, just send the proposed `assert` command. But if you like it, I can follow up.

amtoine commented:
@presidento
thanks a lot for this, it's already something that came up in #8368 and i do think having output on the `std assert` familly of commands is a must-have :partying_face: 

i'll have a look after lunch for sure :wink:

amtoine commented on crates/nu-utils/standard_library/std.nu:
````diff
@@ -8,20 +8,31 @@ def _assert [
 }
 
 # ```nushell
+# >_ let a = 3
````
good catch!

amtoine commented on crates/nu-utils/standard_library/std.nu:
```diff
@@ -8,20 +8,31 @@ def _assert [
-    _assert $cond "condition given to `assert` does not hold"
+export def assert [cond: bool, message?: string] {
+    if $cond { return }
+    let msg_text = (if ($message != null) { $message } else { "No message" })
```
you can use
```rust
$message | default "No message"
```
here :+1: 

i think it's a bit cleaner maybe?

presidento replied:
Good idea, thank you.

sholderbach replied:
I personally dislike the "No message" message. It doesn't really provide guidance and more confusing noise.

presidento replied:
@sholderbach I agree, removed that.

amtoine reviewed (changes requested):
loving it :rocket: 

in order to have these beautiful error messages on `assert eq` and `assert ne` as well, could you modify `_assert` instead of `assert`? :yum:

presidento commented:
Here is the next step. The `_assert` cannot be used, since making errors suppose that it was directly called from user code. (If `assert ne` calls `assert`, it will just show that `$left` and `$right` are not equal.) I extracted the error making step, the only thing which can be common.

I implemented it for `assert eq`, although I think it would be better to use full name, not abbreviation: `assert equal`. For example Nushell uses `str substring`, not `str substr`. Also I think it would better to use `assert not equal` instead of `assert ne`.

I am confused with `assert ne`:

```
>_ 1 != "2"
true

>_ assert ne 1 "2"
Error:
  × left and right operand of `assert eq` should have the same type
   ╭─[myscript.nu:5:1]
 5 │     if not $cond {
 6 │         error make {msg: $msg}
   ·         ─────┬────
   ·              ╰── originates from here
 7 │     }
   ╰────
```

amtoine commented:
@presidento
no need to rebase and force push, it's normal to apply changes as reviews come to a PR :wink: 

and that's not the most convenient for reviewers to checkout your branch and test what has been done, i think :yum:

amtoine commented:
i think the names have been chosen to reflect what is done with `rust`, i.e. `assert`, `assert_eq` and `assert_ne` :+1: 
maybe that should be for another PR?

and about `assert ne` it is simply the exact same thing as `assert eq` but with the boolean equality check being reversed, i.e. `==` becomes a `!=` :wink:

amtoine commented on crates/nu-utils/standard_library/std.nu:
```diff
@@ -7,50 +7,77 @@ def _assert [
 
+def _assertion-error [start, end, label, message?: string] {
+    error make {
+        msg: $"Assertion failed: ($message)",
```
maybe we could only print the `message` when it is given and `Assertion failed.` when it is not?

something like
```diff
-        msg: $"Assertion failed: ($message)",
+        msg: ($message | default "Assertion failed."),
```

amtoine replied:
the examples below would have to be modified though

presidento replied:
These are all assertion errors. I think there can be _additional_ information if there is a message. For example:

```
assert equal $code 301 "The http site is permanently redirected to https one"
```

It it fails, it is an assertion error, and the message helps debugging what was the original idea.

amtoine replied:
yep!

so i'd say that, when a message is given, there is no need to prepend "Assertion failed:" just print the message :yum: 
but when there is no message given, a simple "Assertion failed." might be enough

in the end, it's the `Assertion failed:` when the message is empty that i find strange :open_mouth:

presidento replied:
As you wish.

amtoine commented on crates/nu-utils/standard_library/std.nu:
```diff
@@ -7,50 +7,77 @@ def _assert [
+    let right_end = (metadata $right).span.end
+    if ($left_type != $right_type) {
+        _assertion-error $left_start $right_end $"Different types cannot be equal: ($left_type) <-> ($right_type)." $message
+    }
```
i'd personnally add a few newlines around the `if` statements :yum:

presidento replied:
I happily follow your personal preferences if they are the official ones here. I assume yes, but I am not sure. (I don't want to be asked for a revert if I follow the guidance.)

sholderbach replied:
I see @amtoine is volunteering to write the upcoming `nufmt` to settle all those disputes through technical means :P

amtoine replied:
on my TODO for some time, coming closer and closer lately :relieved:

amtoine replied:
> I happily follow your personal preferences if they are the official ones here. I assume yes, but I am not sure. (I don't want to be asked for a revert if I follow the guidance.)
>
nothing official, but my experience guides me and says it's not good practice to write monolithic blocks of code, maybe except when writing long pipes in shell languages, when the source can be split into several logical pieces :wink:

presidento commented:
In my organization we are allowed to push only perfect commits to master, so we rebase&change commits as long as they are not in the preferred state. If it has no value for this repository, I can leave existing commits as they are and do every change in follow-up commits.

amtoine commented:
> In my organization we are allowed to push only perfect commits to master, so we rebase&change commits as long as they are not in the preferred state. If it has no value for this repository, I can leave existing commits as they are and do every change in follow-up commits.
>
`nushell` squash all the commits and rebase the commit onto the `main` branch.
the effect is that intermediate commits never end up in the `main` branch, but we can still reopen a PR branch and see a particular commit with a bug introduced for instance :blush:

amtoine commented on crates/nu-utils/standard_library/std.nu:
```diff
@@ -1,85 +1,96 @@
+export def "assert ne" [left: any, right: any, message?: string] {
+    let left_start = (metadata $left).span.start
+    let right_end = (metadata $right).span.end
+
```
you're just missing the type check here :+1: 
just to have `assert eq` and `assert ne` be consistent between each other :yum:

presidento replied:
Sorry, I just forgot to update the tests, it was my mistake.

I also want consistency, see https://github.com/nushell/nushell/pull/8405#issuecomment-1464944357. If `$a == $b` then `assert eq $a $b` should pass. Similarly if `$a != $b` then `assert ne $a $b` should pass. I do not agree that `assert ne 1 "2"` should fail.

I can update the tests to assert to the right behaviour or revert changing `assert ne`, then  it is up to you if you want implement the new style for it as well. (Just please document that it is intentionally not consistent with the `!=` operator.)

presidento replied:
I updated the tests.

amtoine reviewed (changes requested):
one extra thing apart from the type check in `assert ne`, the tests do not pass anymore, so there might be something to investigate there to fix them :relieved: 

```bash
nu crates/nu-utils/standard_library/tests.nu
```
now fails

amtoine commented:
not sure what to do with these failing tests :thinking: 

i think `assert eq` and `assert ne` should raise an error when the types of the left and right operands are different but more importantly, with or without this type check, they shoudld definitely be consistent :yum: 

> **Note**
> maybe we need more feedback on this `assert ne`
> should `assert ne` not check the type because things like `1 != "2"` work?
> in that case, the comparison looks ill defined to me because `1` and `"2"` do not even have the same type, so yeah it is `true` that they are not equal, but to me they look not even comparable :open_mouth:

bobhy commented:
It appears that Nushell will let you compare items of different types without complaining about type mismatch.
Therefore, I don't think `assert eq` or `assert ne` should insist that both sides be of the same type.
```nushell
〉1 == 1.0
true
〉true == 1
false
```
The `1 == 1.0` case isn't a crystal clear example of this, but surely `true` is boolean and `1` is not.

bobhy commented:
Due to a race condition, this PR drops out the unit tests for `dirs` that were added earlier today.  Please rebase this on the very latest `main` to pick them up.

This is developer-hostile design, we shouldn't have such race conditions for unrelated sub-modules.  We need a more sophisticated test harness for stdlib so `dirs.nu` could simply provide a `test_dirs.nu` and not have to modify `test.nu`.

presidento commented:
Who is the leader here? As I started to receive contradicting requests, I
need to know whom to follow.

1) May I rebase this branch? If yes, can I squash it into one commit?

2) Can I keep `assert $a != $b` consistent with `assert ne $a $b`?

amtoine commented:
@fdncred @kubouch 

i think we need some help here to decide what to do with `assert eq` and `assert ne` :open_mouth: 

**TL;DR**: should `assert eq 1 "1"` and `assert ne 1 "2"` give errors or something else?

amtoine commented:
@presidento 

i can say for sure @bobhy and i are not members of the core team :+1: 

the standard library is a very recent project with a lot of design being defined right now :open_mouth: 
i hope did not induce too much confusion about who was "in charge" and what had to be done here...

presidento commented:
There is no problem. Sometimes we need to escalate and involve somebody in charge.

@fdncred @kubouch Please include `assert eq 1 1.0` to this topic as well. (At current implementation it gives error, while `1 == 1.0` is true.)

kubouch commented:
So the question is whether the asserts should be type-sensitive, or not? I'm leaning towards mirroring the Nushell's `==`, `!=`, etc., but I don't have a strong preference.

The main point of this PR is to add message to assert, though, right? I would concentrate on that and we can leave the type sensitivity issue to further discussion / another PR. So, I think the PR shouldn't change the way it works currently, just add the custom message.

amtoine commented:
> So the question is whether the asserts should be type-sensitive, or not? I'm leaning towards mirroring the Nushell's `==`, `!=`, etc., but I don't have a strong preference.
> 
> The main point of this PR is to add message to assert, though, right? I would concentrate on that and we can leave the type sensitivity issue to further discussion / another PR. So, I think the PR shouldn't change the way it works currently, just add the custom message.
>
agree with this :+1:

presidento commented:
As you wish. Will do it on Monday.

fdncred commented:
I'd defer to @kubouch's opinion.

presidento commented:
(I reverted the change of behaviour and created additional issues for the mentioned topics. Now this PR is just about the detailed error messages and be able to add custom message to assert methods.)

amtoine reviewed (changes requested):
i wanted to approve because it looked great, but on further inspection, it does not behave as before i think :open_mouth:

@fdncred @kubouch 
i do not want to slow down this PR too much, should i investigate to make sure the behaviour is the same? :yum:

fdncred commented:
By reading the posts here, it seems like it should work before except `Now this PR is just about the detailed error messages and be able to add custom message to assert methods`. If that's not the case, then someone should investigate to try and understand what changed.

amtoine commented:
> assert
> 
:+1: 

> assert eq
>
:+1: 

> # assert ne
> Previously `if types are equal: error` then `if not ($left != $right): error`.
> Now `if $left == $right: error` then `if $type_left != $type_right: error`.
>
this is where i disagree :yum: 

previously we had
```
assert ne [l, r] {
    assert (l.type == r.type)  # error when `l.type != r.type`
    assert (l.val != r.val)
}
```
but now we have
```
assert ne [l, r] {
    if (l.val == r.val): error
    if (l.type != r.type): error
}
```
which does not look the same to me :open_mouth: 

- maybe the definitions of `==` and `!=` in `nushell` do not show the non equivalence
- but as the type check is not after the value check, the two are clearly not equivalent

i hope i'm not becoming blind to what is written here, i've seen so much `assert`s these days :laughing:

amtoine commented:
i'm sorry being that picky here...
i'm just really confused about the logical inversion introduced by this PR :thinking: 

before we had the `_assert` wrapper that, at least to me, made the checks in `assert`, `assert eq` and `assert ne` quite clear.
now all the boolean tests have been inverted and even some of the orders between the tests themselves, to accomodate to the new `_assertion-error`.

presidento commented:
At the end of the day it is your code. If you don't like the style, feel free to "be inspired" with this PR, reject this one and resend a new one with the style you want. I will 👍 it. From functional point of view I am sure they are identical.

amtoine commented:
> At the end of the day it is your code.
>
nope it is absolutely not, i know "my" asserts had a lot of flaws that need to be addressed :+1: 

> From functional point of view I am sure they are identical.
>
i did not check all the workings of `==` and `!=` in `nushell`, so i trust you on this :relieved: 

> If you don't like the style, feel free to "be inspired" with this PR, reject this one and resend a new one with the style you want. I will +1 it.
>
what i'm saying is that the whole logic inversion introduced here is confusing to me, merely to add better message that just require passing the `span` of the condition being tested :thinking: 
i might be the only one confused, but still, i was :wink: 

in the end, "changing the style" might have been better kept in a separate and clearer PR, leaving the "better messages" (and they truely are, don't get me wrong :muscle:) for this one.

@fdncred 
i think we're good :wink:

presidento commented:
For having better error messages, the error-making part must be extracted, since it is the (only) common part. It has an inverse behaviour compared to `_assert`. So keeping the code as it was while extracting the inverse behaviour, well, can be done, but it would made the code more complex. I rewrote the `assert ne` command to use exactly the same checks as the original one had. I hope it makes clear that there were no behaviour changes here.

fdncred commented on crates/nu-utils/standard_library/std.nu:
```diff
@@ -18,88 +18,123 @@ export-env {
+    let left_start = (metadata $left).span.start
+    let right_end = (metadata $right).span.end
+
+    if (($left | describe) == ($right | describe)) {
```
Maybe check for inequality here so you don't have to have a no-op block after if?
```suggestion
    if (($left | describe) != ($right | describe)) {
```

fdncred commented on crates/nu-utils/standard_library/std.nu:
```diff
@@ -18,88 +18,123 @@ export-env {
+    } else {
+        _assertion-error $left_start $right_end $"They have different types: ($left_type) <-> ($right_type)." $message
+    }
+    if ($left != $right) {
```
Same idea here. 

```suggestion
    if ($left == $right) {
```

fdncred commented:
Other than not having no-op blocks, this code seems pretty good to me.

presidento commented:
Done

sholderbach commented:
Let's land this then :)

amtoine commented:
:partying_face: noishe
